paasta_tools/kubernetes/application/controller_wrappers.py

```diff
@@ -188,14 +203,173 @@ def deep_delete(self, kube_client: KubeClient) -> None:
                 )
             )
         self.delete_pod_disruption_budget(kube_client)
+        self.delete_horizontal_pod_autoscaler(kube_client)
+
+    def get_existing_app(self, kube_client: KubeClient):
+        return kube_client.deployments.read_namespaced_deployment(
+            name=self.item.metadata.name, namespace=self.item.metadata.namespace
+        )
 
     def create(self, kube_client: KubeClient) -> None:
         create_deployment(kube_client=kube_client, formatted_deployment=self.item)
         self.ensure_pod_disruption_budget(kube_client)
+        self.sync_horizontal_pod_autoscaler(kube_client)
 
     def update(self, kube_client: KubeClient) -> None:
+        # If autoscaling is enabled, do not update replicas.
+        # In all other cases, replica is set to max(instances, min_instances)
+        if not self.get_soa_config().get("instances"):
+            self.item.spec.replicas = self.get_existing_app(kube_client).spec.replicas
         update_deployment(kube_client=kube_client, formatted_deployment=self.item)
         self.ensure_pod_disruption_budget(kube_client)
+        self.sync_horizontal_pod_autoscaler(kube_client)
+
+    def sync_horizontal_pod_autoscaler(self, kube_client: KubeClient) -> None:
+        """
+        In order for autoscaling to work, there needs to be at least three configurations
+        min_instnace, max_instance, and instance
+        """
+        self.logging.info(
+            "Syncing HPA setting for {self.item.metadata.name}/name in {self.item.metadata.namespace}"
+        )
+        hpa_exists = self.exists_hpa(kube_client)
+        # NO autoscaling
+        if self.get_soa_config().get("instances"):
+            # Remove HPA if autoscaling is disabled
+            if hpa_exists:
+                self.delete_horizontal_pod_autoscaler(kube_client)
+                return
+        # Get autoscaling configurations
+        min_replicas = self.get_soa_config().get("min_instances")
+        max_replicas = self.get_soa_config().get("max_instances")
+        if not min_replicas or not max_replicas:
+            self.logging.error(
+                "min/max_instances are not specified. Autoscaling is not enabled."
+            )
+            return
+        metrics_provider = (
+            self.get_soa_config()
+            .get("autoscaling", {})
+            .get("metrics_provider", "mesos_cpu")
```

Comment:
Here and in several places below, why there is "mesos"?

Reply:
This is the autoscaling config we have in yelpsoa_configs right now. I don't have plans on changing right now.